pkg/verifier/notation/revocationfactory.go

```diff
@@ -14,14 +14,84 @@
 package notation
 
 import (
+	"context"
+	"crypto/x509"
+	"net/http"
+	"sync"
+
 	"github.com/notaryproject/notation-core-go/revocation"
 	corecrl "github.com/notaryproject/notation-core-go/revocation/crl"
+	"github.com/notaryproject/notation-go/dir"
+	"github.com/notaryproject/notation-go/verifier/crl"
+	"github.com/ratify-project/ratify/internal/logger"
 )
 
+// RevocationFactory is an interface that defines methods for creating instances
+// related to revocation. It provides methods to create a new fetcher and a new
+// validator.
 type RevocationFactory interface {
 	// NewFetcher returns a new fetcher instance
 	NewFetcher() (corecrl.Fetcher, error)
 
 	// NewValidator returns a new validator instance
 	NewValidator(revocation.Options) (revocation.Validator, error)
 }
+
+// NewFetcher returns a new fetcher instance
+func NewFetcher(httpClient *http.Client, cacheRoot string) (corecrl.Fetcher, error) {
+	crlFetcher, err := corecrl.NewHTTPFetcher(httpClient)
+	if err != nil {
+		return nil, err
+	}
+	crlFetcher.Cache, err = newFileCache(cacheRoot)
+	if err != nil {
+		return nil, err
+	}
+	return crlFetcher, nil
+}
+
+// SupportCRL checks if the certificate supports CRL
+func SupportCRL(cert *x509.Certificate) bool {
+	return cert != nil && len(cert.CRLDistributionPoints) > 0
+}
+
+// cacheCRL caches the Certificate Revocation Lists (CRLs) for the given certificates using the provided CRL fetcher.
+// It logs a warning if fetching the CRL fails but does not return an error to ensure the process is not blocked.
+func CacheCRL(ctx context.Context, certs []*x509.Certificate, fetcher corecrl.Fetcher) {
+	if fetcher == nil {
+		logger.GetLogger(ctx, logOpt).Warn("CRL fetcher is nil")
+		return
+	}
+	var wg sync.WaitGroup
+	for _, cert := range certs {
+		if !SupportCRL(cert) {
+			return
+		}
+		cacheCertificateCRL(ctx, cert.CRLDistributionPoints, fetcher, &wg)
+	}
+	wg.Wait()
+}
+
+func cacheCertificateCRL(ctx context.Context, crlURLs []string, crlFetcher corecrl.Fetcher, wg *sync.WaitGroup) {
```

Comment:
I would recommend to add unit test covering the multiple-goroutine scenario.

Reply:
Sure thing.

Reply:
Updated